cekit/cache/artifact.py

```diff
@@ -37,43 +42,35 @@ def add(self, artifact):
         if not set(SUPPORTED_HASH_ALGORITHMS).intersection(artifact):
             raise ValueError('Cannot cache Artifact without checksum')
 
-        artifact_id = str(uuid.uuid4())
-
-        for alg in SUPPORTED_HASH_ALGORITHMS:
-            if alg in artifact:
-                self._index_artifact(artifact, artifact_id, alg, artifact[alg])
-            else:
-                self._index_artifact(artifact, artifact_id, alg)
+        if self.is_cached(artifact):
+            raise CekitError('Artifact is already cached')
 
-        return artifact_id
+        artifact_id = str(uuid.uuid4())
 
-    def _index_artifact(self, artifact, artifact_id, alg=None, chksum=None):
         artifact_file = os.path.expanduser(os.path.join(self._cache_dir, artifact_id))
-
         if not os.path.exists(artifact_file):
             artifact.guarded_copy(artifact_file)
 
-        if chksum:
-            if not check_sum(artifact_file, alg, chksum):
-                raise CekitError('Artifact contains invalid checksum!')
-        else:
-            chksum = get_sum(artifact_file, alg)
-
-        cache = self._get_cache()
-        if artifact_id not in cache:
-            cache[artifact_id] = {}
+        cache_entry = {'names': [artifact['name']],
+                       'cached_path': artifact_file}
 
-        cache[artifact_id].update({'names': [artifact['name']],
-                                   alg: chksum,
-                                   'cached_path': artifact_file})
+        for alg in SUPPORTED_HASH_ALGORITHMS:
+            if alg in artifact:
+                if not check_sum(artifact_file, alg, artifact[alg]):
+                    raise CekitError('Artifact contains invalid checksum!')
+                chksum = artifact[alg]
+            else:
+                chksum = get_sum(artifact_file, alg)
+            cache_entry.update({alg: chksum})
 
-        self._update_cache(cache)
+        self._update_cache(cache_entry, artifact_id)
+        return artifact_id
 
     def delete(self, uuid):
         cache = self._get_cache()
         artifact = cache.pop(uuid)
         os.remove(os.path.expanduser(artifact['cached_path']))
```

Comment:
Why not just?

```
os.remove(os.path.join(self._cache_dir, uuid))
```

Tne you would not need to to even get it from cache above.

Reply:
I need to get it from cache above as if its not cached I'll print error